src/blockchain_db/sqlite/db_sqlite.cpp

```diff
@@ -59,38 +61,60 @@ BlockchainSQLite::BlockchainSQLite(cryptonote::network_type nettype, fs::path db
   }
 }
 
-
 void BlockchainSQLite::create_schema() {
   LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__);
 
 	db.exec(R"(
-CREATE TABLE batched_payments (
+CREATE TABLE batched_payments_accrued (
+    address VARCHAR NOT NULL,
+    amount BIGINT NOT NULL,
+    PRIMARY KEY(address),
+    CHECK(amount >= 0)
+);
+
+CREATE TABLE batched_payments_paid (
     address VARCHAR NOT NULL,
     amount BIGINT NOT NULL,
-    height_earned BIGINT NOT NULL,
     height_paid BIGINT,
-    PRIMARY KEY(address, height_earned),
-    CHECK(amount > 0)
+    PRIMARY KEY(address, height_paid),
+    CHECK(amount >= 0)
 );
-CREATE VIEW accrued_rewards AS
-    SELECT 
-        address,
-        SUM(amount) as amount,
-        MIN(height_earned) as height
-    FROM batched_payments 
-    WHERE height_paid IS NULL 
-    GROUP BY address;
+
+CREATE INDEX batched_payments_paid_height_idx ON batched_payments_paid (height_paid);
 
 CREATE TABLE batch_db_info (
-    height BIGINT NOT NULL
+    height BIGINT NOT NULL,
+    fire_trigger INT NOT NULL
 );
 
-INSERT INTO batch_db_info (height) VALUES (0);
+INSERT INTO batch_db_info (height, fire_trigger) VALUES (0, 1);
 
 CREATE TRIGGER batch_payments_prune_paid
-AFTER UPDATE ON batch_db_info FOR EACH ROW
+AFTER UPDATE ON batch_db_info
+FOR EACH ROW
+BEGIN
+    UPDATE batch_db_info SET fire_trigger = 0;
+    DELETE FROM batched_payments_paid WHERE height_paid < (NEW.height - 10000);
+    UPDATE batch_db_info SET fire_trigger = 1;
+END;
```

Comment:
This `fire_trigger` thing is a little bit ugly.  An alternative to consider is to use a view as indirection with `INSTEAD OF` triggers on the view--so select, insert, and delete from the view normally, and work with the underlying raw table if you need to bypass the triggers.

I've also added another tweak here to make queries fail if you attempt to add a paid record for a wallet that doesn't exist (currently this is just silently not doing anything).

```SQL
CREATE TABLE batched_payments_accrued (
    address VARCHAR NOT NULL,
    amount BIGINT NOT NULL,
    PRIMARY KEY(address),
    CHECK(amount >= 0)
);

CREATE TRIGGER batched_payments_delete_empty
AFTER UPDATE ON batched_payments_accrued
FOR EACH ROW WHEN NEW.amount = 0
BEGIN
    DELETE FROM batched_payments_accrued WHERE address = NEW.address;
END;

CREATE TABLE batched_payments_raw (
    address VARCHAR NOT NULL,
    amount BIGINT NOT NULL,
    height_paid BIGINT,
    PRIMARY KEY(address, height_paid),
    CHECK(amount >= 0)
);

CREATE INDEX batched_payments_raw_paid_height_idx ON batched_payments_raw (height_paid);

CREATE TABLE batch_db_info (
    height BIGINT NOT NULL
);
INSERT INTO batch_db_info (height) VALUES (0);

CREATE TRIGGER batch_payments_prune
AFTER UPDATE ON batch_db_info
FOR EACH ROW
BEGIN
    DELETE FROM batched_payments_raw WHERE height_paid < (NEW.height - 10000);
END;


CREATE VIEW batched_payments_paid AS SELECT * FROM batched_payments_raw;

CREATE TRIGGER make_payment INSTEAD OF INSERT ON batched_payments_paid
FOR EACH ROW
BEGIN
    UPDATE batched_payments_accrued SET amount = amount - NEW.amount
        WHERE address = NEW.address;
    SELECT RAISE(ABORT, 'Address not found') WHERE changes() = 0;
    INSERT INTO batched_payments_raw (address, amount, height_paid)
        VALUES (NEW.address, NEW.amount, NEW.height_paid);
END;

CREATE TRIGGER rollback_payment
INSTEAD OF DELETE ON batched_payments_paid
FOR EACH ROW
BEGIN
    DELETE FROM batched_payments_raw
        WHERE address = OLD.address AND height_paid = OLD.height_paid;
    INSERT INTO batched_payments_accrued
        (address, amount) VALUES (OLD.address, OLD.amount)
        ON CONFLICT (address) DO UPDATE SET amount = amount + excluded.amount;
END;
```

and some examples:

```
sqlite> insert into batched_payments_accrued (address, amount) values ("a1", 123) on conflict (address) do update set amount = amount + excluded.amount;
sqlite> insert into batched_payments_accrued (address, amount) values ("a1", 123) on conflict (address) do update set amount = amount + excluded.amount;
sqlite> insert into batched_payments_accrued (address, amount) values ("a1", 123) on conflict (address) do update set amount = amount + excluded.amount;
sqlite> insert into batched_payments_accrued (address, amount) values ("a1", 123) on conflict (address) do update set amount = amount + excluded.amount;
sqlite> insert into batched_payments_accrued (address, amount) values ("a1", 123) on conflict (address) do update set amount = amount + excluded.amount;
sqlite> insert into batched_payments_accrued (address, amount) values ("a1", 123) on conflict (address) do update set amount = amount + excluded.amount;
sqlite> insert into batched_payments_accrued (address, amount) values ("a1", 123) on conflict (address) do update set amount = amount + excluded.amount;
sqlite> insert into batched_payments_accrued (address, amount) values ("a1", 123) on conflict (address) do update set amount = amount + excluded.amount;
sqlite> insert into batched_payments_accrued (address, amount) values ("a1", 123) on conflict (address) do update set amount = amount + excluded.amount;
sqlite> insert into batched_payments_accrued (address, amount) values ("a1", 123) on conflict (address) do update set amount = amount + excluded.amount;
sqlite> select * from batched_payments_accrued;
address  amount
-------  ------
a1       1230  
sqlite> insert into batched_payments_paid (address, amount, height_paid) values ("a1", 1230, 45678);
sqlite> select * from batched_payments_accrued;
sqlite> select * from batched_payments_paid;
address  amount  height_paid
-------  ------  -----------
a1       1230    45678      
sqlite> delete from batched_payments_paid WHERE address = "a1" AND height_paid = 45679;
sqlite> select * from batched_payments_paid;
address  amount  height_paid
-------  ------  -----------
a1       1230    45678      
sqlite> delete from batched_payments_paid WHERE address = "a1" AND height_paid = 45678;
sqlite> select * from batched_payments_paid;
sqlite> select * from batched_payments_accrued;
address  amount
-------  ------
a1       1230  
sqlite> insert into batched_payments_paid (address, amount, height_paid) values ("a1", 1230, 45678);
sqlite> update batch_db_info set height = 88888;
sqlite> select * from batched_payments_accrued;
sqlite> select * from batched_payments_paid;
sqlite> insert into batched_payments_paid (address, amount, height_paid) values ("a1", 1230, 45678);
Error: stepping, Address not found (19)
sqlite> 
```

I think all the current code should work fine with that change, except for the test code that sets `fire_trigger`: that code can be changed to just work on the `_raw` table instead of the view.